thor/src/keys.rs

```diff
@@ -9,6 +8,19 @@ use ecdsa_fun::{
 use sha2::Sha256;
 use std::fmt;
 
+pub type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Debug, Clone, Copy, thiserror::Error)]
+pub enum Error {
```

Comment:
🔧 I like the idea of using explicit types for errors, but I would advise against using an enum to model _all_ the errors in a module. For example when you use the `Error` enum in the `verify_revocation_secret_key` method below you're saying that it could fail by the appropriate `WrongRevocationSecreKey` error _and_ all the other variants, which is semantically incorrect.

I think it's better to model the errors of each function separately. In some cases that will require an enum, but other times you will just need a struct.

Reply:
I understand your point and I have to think about what are the best practices in the Rust ecosystem to find a better solution.

Looking at good quality Rust crates, they would usually have one `Error` enum at the crate's root and possibly some more specific `Error` enums in some specialist sub-modules.

What you are saying is that we could leverage more the type systems by using unit struct as Error of functions that can only fail one way.

However I don't agree with this statement:

>  you're saying that it could fail by the appropriate WrongRevocationSecreKey error and all the other variants, which is semantically incorrect.

IMHO, in any crate where they decide to have an `Error` enum at the crate's root, they don't say that all functions in the crate can fail with all variants of the `Error` enum, they just make it easier for the consumer to handle errors by providing a unique `Error` type to handle.

And again, sometimes there are other specialised `Error` enum in the sub-modules which one way to say "Most consumers would not use functions in this module so they should not have to handle error variants that can only be returned by those functions, hence, let's define a new Error type".

I also understand that the choice of using unit structs was driven by providing nice errors messages using a combination of `thiserrors` and `anyhow`. But looking closer to it:

```
fn verify_revocation_secret_key(&self, secret_key: &RevocationSecretKey) -> Result<(), WrongRevocationSecreKey> {
```

Shouldn't it be?

```
fn is_revocation_secret_key_correct(&self, secret_key: &RevocationSecretKey) -> bool {
```

Using a unit struct as a the `Error` hand of a `Result`, for which the `Ok` hand is the unit type sounds like an abuse of the `Result` type.

This function assesses whether the key is valid/correct.
The call site can then decide what to do with this information. Whether it's bubbling up an error, banning the peer, restarting the protocol, etc.

Reply:
The function I pointed to was just an example. You could definitely return a `bool` instead of a `Result` if you want, I'm not fussed about that.

> they just make it easier for the consumer to handle errors by providing a unique Error type to handle

If you use an `Error` enum as part of the return type of a function, if said enum contains variants that are not related to the function, a consumer will have to handle all the semantically invalid variants when handling the error, no?